kernel/src/scan/mod.rs

```diff
@@ -161,28 +162,23 @@ impl Scan {
         let parquet_handler = engine_interface.get_parquet_handler();
 
         let partition_columns = &self.snapshot.metadata().partition_columns;
-        let read_schema = Arc::new(StructType::new(
-            self.schema()
-                .fields()
-                .filter(|f| !partition_columns.contains(f.name()))
-                .cloned()
-                .collect::<Vec<_>>(),
-        ));
-
-        let mut partition_fields = partition_columns
-            .iter()
-            .map(|column| {
-                self.schema()
-                    .field(column)
-                    .ok_or(Error::generic("Unexpected partition column"))
-            })
-            .collect::<DeltaResult<Vec<_>>>()?;
-        partition_fields.reverse();
-
-        let select_fields = read_schema
-            .fields()
-            .map(|f| Expression::column(f.name()))
-            .collect_vec();
+        for partition_column in partition_columns.iter() {
+            if !self.schema().contains(partition_column) {
+                return Err(Error::Generic(format!(
+                    "Unexpected partition column {partition_column}"
+                )));
+            }
+        }
+        let mut partition_fields = vec![];
```

Comment:
maybe `Vec::with_capacity` if you know the lengths ahead of time to avoid an allocation.

Reply:
yeah thanks. ended up doing that in final version